TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/KafkaPrinter.java

```diff
@@ -126,26 +192,61 @@ public static void main(String[] args) {
         }
     }
 
+    static boolean checkAllRecordsCompleted(Collection<PartitionTracker> trackers) {
+        for (PartitionTracker tracker : trackers) {
+            if (tracker.currentRecordCount < tracker.recordLimit) {
+                return false;
+            }
+        }
+        return true;
+    }
+
     static void pipeRecordsToProtoBufDelimited(
-        Consumer<String, byte[]> kafkaConsumer,
-                                               java.util.function.Consumer<Stream<byte[]>> binaryReceiver) {
-        while (true) {
-            processNextChunkOfKafkaEvents(kafkaConsumer, binaryReceiver);
+        Consumer<String, byte[]> kafkaConsumer, java.util.function.Consumer<Stream<ConsumerRecord<String, byte[]>>> binaryReceiver,
+        long timeoutSeconds, Map<Partition, PartitionTracker> capturedRecords) {
+
+        long endTime = System.currentTimeMillis() + (timeoutSeconds * 1000);
+        boolean continueCapture = true;
+        while (continueCapture) {
+            if (!capturedRecords.isEmpty() && checkAllRecordsCompleted(capturedRecords.values())) {
+                log.info("All partition limits have been met, stopping Kafka polls");
+                continueCapture = false;
+            }
+            else if (timeoutSeconds > 0 && System.currentTimeMillis() >= endTime) {
+                log.warn("Specified timeout of {} seconds has been breached, stopping Kafka polls", timeoutSeconds);
+                continueCapture = false;
+            }
+            else {
+                for (PartitionTracker pt : capturedRecords.values()) {
+                    log.debug("Tracker is at {} records for limit {}", pt.currentRecordCount, pt.recordLimit);
+                }
+                processNextChunkOfKafkaEvents(kafkaConsumer, binaryReceiver);
+            }
         }
     }
 
-    static void processNextChunkOfKafkaEvents(Consumer<String, byte[]> kafkaConsumer, java.util.function.Consumer<Stream<byte[]>> binaryReceiver) {
+    static void processNextChunkOfKafkaEvents(Consumer<String, byte[]> kafkaConsumer, java.util.function.Consumer<Stream<ConsumerRecord<String, byte[]>>> binaryReceiver) {
         var records = kafkaConsumer.poll(CONSUMER_POLL_TIMEOUT);
-        binaryReceiver.accept(StreamSupport.stream(records.spliterator(), false)
-                .map(ConsumerRecord::value));
+        binaryReceiver.accept(StreamSupport.stream(records.spliterator(), false));
```

Comment:
do you want to flatten records from across different partitions into just one stream?  See my comments about preserving partition boundaries.

Reply:
Have added default to preserve partition boundaries and create separate files